roll/ast.py

```diff
@@ -0,0 +1,685 @@
+
+    @trace
+    def hash_vars(self, counter, map):
+        pass
+
+    def __str__(self):
+        return str(self.__value)
+
+    @property
+    def pure(self):
+        return self.__value
+
+    def __eq__(self, other):
+        return isinstance(other, TokenNumber) and self.__value == other.__value
+
+
+class TokenString(Token):
+    def __init__(self, value):
+        self.__value = value
+        logging.debug(self.__class__.__name__, self)
+
+    @trace
+    def reduce(self, env, counter):
+        return self
+
+    def substitute(self, _):
+        return TokenString(self.__value)
+
+    def type(self, _):
+        return TypeString
+
+    @trace
+    def hash_vars(self, counter, map):
+        pass
+
+    def __str__(self):
+        unescaped = '"'
+        escaped = '"'
+        return f'"{self.__value.replace(unescaped, escaped)}"'
+
+    @property
+    def pure(self):
+        return self.__value
+
+    def __eq__(self, other):
+        return isinstance(other, TokenString) and self.__value == other.__value
+
+
+class TokenRoll(Token):
+    def __init__(self, count, sides):
+        self.count = count
+        self.sides = sides
+        logging.debug(self.__class__.__name__, self)
+
+    @trace
+    def reduce(self, env, counter):
+        sides = self.sides.reduce(env, counter).pure
+        if sides == 0:
+            raise rollerr.ZeroDiceSidesError(env.trace)
+        if sides < 0:
+            raise rollerr.NegativeDiceSidesError(env.trace, sides)
+        if int(sides) != sides:
+            raise rollerr.FloatingPointDiceSidesError(env.trace, sides)
+        count = self.count.reduce(env, counter).pure
+        if count == 0:
+            raise rollerr.ZeroDiceCountError(env.trace)
+        if count < 0:
+            raise rollerr.NegativeDiceCountError(env.trace, count)
+        if int(count) != count:
+            raise rollerr.FloatingPointDiceCountError(env.trace, count)
+        if sides == 1:
+            return TokenNumber(count)
+        if count > MAX_ROLLS:
+            raise rollerr.ExcessiveDiceRollsError(env.trace)
+        return TokenNumber(sum(random.choices(range(1, sides + 1), k=count)))
+
+    def substitute(self, old_to_new):
+        return TokenRoll(
+            self.count.substitute(old_to_new), self.sides.substitute(old_to_new)
+        )
+
+    def type(self, env):
+        # if count.type(env) != Number or sides.type(env) != Number:
+        #     raise rollerr.TypeError(env.trace, "Expecting ")
+        return TypeNumber
+
+    @trace
+    def hash_vars(self, counter, map):
+        self.count.hash_vars(counter, map)
+        self.sides.hash_vars(counter, map)
+
+    def __str__(self):
+        count = str(self.count)
+        sides = str(self.sides)
+        if not isinstance(self.count, TokenNumber):
+            count = f"({count})"
+        if not isinstance(self.sides, TokenNumber):
+            sides = f"({sides})"
+        return f"{count}d{sides}"
+
+
+class TokenVariable(Token):

```

Comment:
Doesn't implement `pure` - intentional?

Reply:
`TokenLet`, `TokenFunction`, and `TokenApplication` also do not implement `pure`.

Reply:
Also `TokenOperator`, `TokenTernary`, `TokenCase`, and `Program`.